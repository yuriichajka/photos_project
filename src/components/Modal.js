import React, {useState} from "react";
import styled from "styled-components";
import {MdClose} from 'react-icons/md'

const Background = styled.div`
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.8);
    position: fixed;
    display: flex;
    justify-content: center;
    align-items: center;
`
const ModalWrapper = styled.div`
    width: 800px;
    height: 500px;
    box-shadow: 0 5px 16px rgba(0, 0, 0, 0.2);
    color: #000;
    background: #fff;
    display: grid;
    grid-template-columns: 2fr 1fr;
    position: relative;
    z-index: 10;
    border-radius: 10px;
`

const ModalContent = styled.div`
    display: flex;
    flex-direction: column;
    margin-top: 40px;
    align-items: center;
    line-height: 1.8;
    color: #141414;
  
  button {
    padding: 10px 24px;
    background: #141414;
    color: #fff;
    border: none;
    width: 75%;
    margin-top: 10px;
  }
  
  input {
    width: 75%;
    margin: 10px 0;
    padding: 5px;
  }
  
  img {
    margin-bottom: 20px;
  }
`

const CloseModalButton = styled(MdClose)`
    cursor: pointer;
    position: absolute;
    top: 20px;
    right: 20px;
    width: 32px;
    height: 32px;
    padding: 0;
    z-index: 10;
`

const Comments = styled.div`
    margin-top: 40px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  
  p {
    text-align: left;
    margin-top: 10px;
  }
  
  div {
    text-align: left;
    color: darkgray;
  }
`


export default function Modal({showModal, setShowModal}) {

    let [name, setName] = useState('')
    let [comm, setComm] = useState('')

    const handleSubmit = (e) => {
        e.preventDefault()

        if (!name || !comm) return

        console.log(name,comm);
    }

    return (
        <>
            {showModal ? (
                <Background>
                    <ModalWrapper showModal={showModal}>
                        <ModalContent>
                            <img src='https://picsum.photos/id/240/300/200' alt="pic" width='400px'/>
                            <form onSubmit={handleSubmit}>
                                <input type="text" value={name} onChange={({target: {value}}) => setName(value)} placeholder='Ваше имя'/>
                                <input type="text" value={comm} onChange={({target: {value}}) => setComm(value)} placeholder='Ваш комментарий'/>
                                <button disabled={!name || !comm}>Оставить отзыв</button>
                            </form>
                        </ModalContent>
                        <CloseModalButton aria-label='Close modal' onClick={() => setShowModal(prev => !prev)}/>
                        <Comments>
                            <div>Lorem ipsum.</div>
                            <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Magni, saepe.</p>
                            <br/>
                            <div>Lorem ipsum.</div>
                            <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Magni, saepe.</p>
                        </Comments>
                    </ModalWrapper>
                </Background>
            ) : null}
        </>
    );
};