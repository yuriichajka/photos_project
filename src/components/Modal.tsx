import React, {useContext, useEffect, useState} from 'react';
import { Col, Form, Row, Button } from 'react-bootstrap';
import { useParams } from 'react-router-dom';
import { Formik } from 'formik';
import * as yup from 'yup';

import { getPhoto } from '../api';

import {
    Background, ModalWrapper,
    CloseModalButton, ModalContainer,
    Comments, Comment, ModalImg
} from '../styled/styledModal';
import { NAME_VAL, COMMENT_VAL } from '../validators';
import { IContext, ModalContext } from '../App';

export interface IImages {
    id?: number
    url?: string
}

const Modal: React.FC = () => {

    const modalContext = useContext<IContext>(ModalContext)

    const { showModal, toggleModal } = modalContext;

    const validations = yup.object().shape({
        name: NAME_VAL,
        comment: COMMENT_VAL
    });

    const { id } = useParams<any>();

    let [modalPhoto, setModalPhoto] = useState<IImages>({});
    useEffect(() => {
        getPhoto(id).then(value => setModalPhoto(value.data))
    }, [id]);

    return (
        <>
            { showModal && (
                    <Background>
                        <ModalWrapper showModal={ showModal }>
                            <ModalContainer>
                               <Row>
                                   <Col>
                                       <ModalImg loading="lazy" src={ modalPhoto.url } width='400px' height='300px'/>
                                       <Formik
                                           initialValues={{
                                                name: '',
                                                comment: ''
                                           }}
                                           validateOnBlur
                                           onSubmit={ (values) => { console.log(values) }}
                                           validationSchema={ validations }
                                       >
                                           {({ values,
                                                 errors,
                                                 touched,
                                                 handleChange,
                                                 handleBlur,
                                                 isValid,
                                                 handleSubmit,
                                                 dirty}) => (
                                               <Form className="d-grid gap-2">
                                                   <Form.Control
                                                       type={ `text` }
                                                       name={ `name` }
                                                       onChange={ handleChange }
                                                       onBlur={ handleBlur }
                                                       value={ values.name }
                                                       placeholder="Name"
                                                   />
                                                   { touched.name && errors.name && <p>{errors.name}</p> }
                                                   <Form.Control
                                                       placeholder="Comment"
                                                       type={ `comment` }
                                                       name={ `comment` }
                                                       onChange={ handleChange }
                                                       onBlur={ handleBlur }
                                                       value={ values.comment }
                                                   />
                                                   { touched.comment && errors.comment && <p>{ errors.comment }</p> }
                                                   <Button
                                                       size="lg"
                                                       style={{ marginBottom: '30px' }}
                                                       disabled={ !isValid && !dirty }
                                                       onClick={ handleSubmit }
                                                       type={ `submit` }
                                                   >Post</Button>
                                               </Form>
                                           )}
                                       </Formik>
                                   </Col>
                                   <Col>
                                       <Comments>
                                           <Comment>
                                               <p className="name">Johnny First</p>
                                               <p className="comment">
                                                   Lorem ipsum dolor sit amet,
                                                   consectetur adipisicing elite.
                                                   Quisquam, reprehenderit.
                                               </p>
                                           </Comment>
                                           <Comment>
                                               <p className="name">Johnny First</p>
                                               <p className="comment">
                                                   Lorem ipsum dolor sit amet,
                                                   consectetur adipisicing elite.
                                                   Quisquam, reprehenderit.</p>
                                           </Comment>
                                       </Comments>
                                   </Col>
                               </Row>
                            </ModalContainer>
                            <CloseModalButton arial-label='Close modal' onClick={ toggleModal }/>
                        </ModalWrapper>
                    </Background>
            )}
        </>
    )
}

export default Modal;
