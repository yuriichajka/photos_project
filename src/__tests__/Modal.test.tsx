import React from 'react';

import Modal from '../components/Modal';

import renderer from 'react-test-renderer'
import {mount, shallow, render} from 'enzyme';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        myRoute: 'some-route-value',
    }),
}));

let wrapper;

describe('Modal snapshot',  () => {
    it('should add a snapshot',  () => {
        wrapper = renderer.create(<Modal />).toJSON();
        expect(wrapper).toMatchSnapshot();
    });
});
describe("Correct render", () => {
    beforeEach(() => {
        // @ts-ignore
        wrapper = mount(<Modal />);
    });

    it('should render something',  () => {
        expect(wrapper).not.toBeNull();
    });

    it('should contain inputs',  () => {
        expect(wrapper.find('input')).toHaveLength(2);
    });

    it('should contain button',  () => {
        expect(wrapper.find('button')).toHaveLength(1);
    });
});

    describe('dad', () => {
        it('should ',  () => {
            const mockCallBack = jest.fn();
            // @ts-ignore
            const component = mount(<Modal onClick={mockCallBack} />);
            component.find('button#test').simulate('click');
            expect(mockCallBack).toBeCalledTimes(1);
        });

        // it('should ',  () => {
        //     const mockCallBack = jest.fn();
        //     // @ts-ignore
        //     const component = mount(<Modal onChange={mockCallBack} />);
        //     expect(mockCallBack.mock.calls.length).toEqual(0);
        //     component.find('input#form1').simulate('click');
        //     expect(mockCallBack.mock.calls.length).toEqual(1);
        // });
    })
