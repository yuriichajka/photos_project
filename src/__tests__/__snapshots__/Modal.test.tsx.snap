// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapshot should add a snapshot 1`] = `
<div
  className="sc-bdnxRM kjnLZx"
>
  <div
    className="sc-dlnjwi hWgIVS"
  >
    <div
      className="sc-hKFxyN hAYNwM"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <img
            className="sc-gKAaRy cKygeM"
            height="300px"
            loading="lazy"
            width="400px"
          />
          <form
            className="d-grid gap-2 commForm"
          >
            <input
              className="form-control"
              id="form1"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />
            <input
              className="form-control"
              id="form2"
              name="comment"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Comment"
              type="comment"
              value=""
            />
            <button
              className="btn btn-primary btn-lg"
              disabled={false}
              id="test"
              onClick={[Function]}
              style={
                Object {
                  "marginBottom": "30px",
                }
              }
              type="submit"
            >
              Post
            </button>
          </form>
        </div>
        <div
          className="col"
        >
          <div
            className="sc-eCApnc iSYAyb"
          >
            <div
              className="sc-jSFjdj ffbhEk"
            >
              <p
                className="name"
              >
                Johnny First
              </p>
              <p
                className="comment"
              >
                Lorem ipsum dolor sit amet, consectetur adipisicing elite. Quisquam, reprehenderit.
              </p>
            </div>
            <div
              className="sc-jSFjdj ffbhEk"
            >
              <p
                className="name"
              >
                Johnny First
              </p>
              <p
                className="comment"
              >
                Lorem ipsum dolor sit amet, consectetur adipisicing elite. Quisquam, reprehenderit.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <svg
      arial-label="Close modal"
      className="sc-gtsrHT fDZzzA"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
</div>
`;
